api-server/src/server/boot/challenge.js

```diff
@@ -82,6 +95,328 @@ export default async function bootChallenge(app, done) {
+  const completedDate = Date.now();
+  let completedChallenge = body;
+
+  let finalChallenge;
+  const $push = {},
+    $set = {},
+    $pull = {};
+  const {
+    completedChallenges = [],
+    needsModeration = false,
+    savedChallenges = []
+  } = user;
+
+  const oldIndex = completedChallenges.findIndex(
+    ({ id: challengeId }) => challengeId === id
+  );
+
+  const alreadyCompleted = oldIndex !== -1;
+  const oldChallenge = alreadyCompleted ? completedChallenges[oldIndex] : null;
+
+  if (alreadyCompleted) {
+    finalChallenge = {
+      ...completedChallenge,
+      completedDate: oldChallenge.completedDate
+    };
+    $set[`completedChallenges.${oldIndex}`] = finalChallenge;
+  } else {
+    finalChallenge = {
+      ...completedChallenge
+    };
+    $push.progressTimestamps = completedDate;
+    $push.completedChallenges = finalChallenge;
+  }
+
+  if (savableChallenges.includes(id)) {
+    const challengeToSave = {
+      id,
+      lastSavedDate: completedDate,
+      files: files?.map(file =>
+        pick(file, ['contents', 'key', 'name', 'ext', 'history'])
+      )
+    };
+
+    const savedIndex = savedChallenges.findIndex(
+      ({ id: challengeId }) => challengeId === id
+    );
+
+    if (savedIndex >= 0) {
+      $set[`savedChallenges.${savedIndex}`] = challengeToSave;
+      savedChallenges[savedIndex] = challengeToSave;
+    } else {
+      $push.savedChallenges = challengeToSave;
+      savedChallenges.push(challengeToSave);
+    }
+  }
+
+  // remove from partiallyCompleted on submit
+  $pull.partiallyCompletedChallenges = { id };
+
+  if (needsModeration) $set.needsModeration = true;
+
+  const updateData = {};
+  if (!isEmpty($set)) updateData.$set = $set;
+  if (!isEmpty($push)) updateData.$push = $push;
+  if (!isEmpty($pull)) updateData.$pull = $pull;
+
+  // ----------------------------------------------
+
+  const points = alreadyCompleted ? user.points : user.points + 1;
+  const updatePromise = new Promise((resolve, reject) =>
+    user.updateAttributes(updateData, (err, instance) => {
+      if (err) {
+        return reject(err);
+      }
+      return resolve(instance);
+    })
+  );
+  return Observable.fromPromise(updatePromise)
+    .map(() => {
+      return res.json({
+        completedChallenges: updateData.$push.completedChallenges,
+        points,
+        savedChallenges
+      });
+    })
+    .subscribe(() => {}, next);
+}
+
+export const expectedProjectStructures = {
+  [challengeTypes.multifileCertProject]: [
+    'files',
+    'files.contents',
+    'files.ext',
+    'files.history',
+    'files.key',
+    'files.name',
+    'id'
+  ],
+  [challengeTypes.bonfire]: [
+    'files',
+    'files.contents',
+    'files.ext',
+    'files.history',
+    'files.key',
+    'files.name',
+    'id'
+  ],
+  [challengeTypes.zipline]: ['id', 'solution'],
+  [challengeTypes.backEndProject]: ['id', 'solution', 'githubLink'],
+  [challengeTypes.pythonProject]: ['id', 'solution', 'githubLink'],
```

Comment:
This seems to break backend and python project submissions - I tried to submit, and it didn't work cause I didn't include a github link. That's always been optional.